state(followers): move (email-)followers to redux

flootr opened the pull request:
#### Changes proposed in this Pull Request

* Migrate `followers` and `email-followers` flux store to redux (using a pre-existing reducer)

#### Testing instructions

1. Go to the people section of a site and choose _Followers_
2. Make sure the list of followers shows up
3. If you scroll down on a page with more than 100 followers you should see another request with incremented `page` param
4. If you switch sites, make sure request params are reset and the list loads fine
5. Try to remove followers, are they deleted from the list and don't appear again on a reload?

Follow the steps above for _Email Followers_ as well.

Note: I left some comments where I would like additional feedback.

Fixes #24218

flootr committed:
wip: move (email-)followers to redux

flootr committed:
reset followers list components on site change

flootr committed:
remove unused follower-data component

flootr committed:
followers: use removeFollower thunk instead of flux action

flootr committed:
remove (email-)followers flux store

flootr committed:
query-followers: implement refresh functionality

flootr committed:
followers-list: fetch 100 entities instead of 50

flootr committed:
query-follower: simplify refresh logic

flootr commented on client/my-sites/people/main.jsx:
```diff
@@ -33,9 +33,16 @@ class People extends React.Component {
 				return <TeamList site={ site } search={ search } />;
 			case 'followers':
-				return <FollowersList site={ site } />;
+				return <FollowersList key={ `people-followers-${ site.ID }` } site={ site } />;
```
This makes sure we get a fresh list component (and don't mess with state updates) in case a user changes their site (opposed to having custom logic in the component itself which listens for changes to `site.id`). I'm not sure though whether this counts as _hacky_ way to accomplish this.

tyxla replied:
🤔  If the user changes their site, `site` prop will change and a new `site` prop will be passed to `FollowersList` anyway. So I'm not sure I understand - can you elaborate?

flootr replied:
We handle internal state (`page`) which won't reset if a user changes their site so if a user fetched a few pages on one site and changes to another site it'd reuse that `page` state which we don't want. Instead of resetting that manually inside the component we just mount a fresh instance which will then get the initial state. I'm not sure if that's a good approach but I've seen it in a few places and it _feels_ a little bit cleaner.

looping in @jsnajdr for additional feedback on this approach

tyxla replied:
This actually makes a lot of sense. Thanks for elaborating, I think it's a solid approach and we've done it before.

If could further nitpick, let's add a comment explaining why we're adding a `key` prop.

flootr replied:
added a comment via 3456f86

flootr commented on client/components/data/query-followers/index.jsx:
```diff
@@ -0,0 +1,50 @@
+	interval = null;
+
+	componentDidMount() {
+		this.props.fetchFollowers( this.props.query );
```
This is different to the original flux implementation where we only fetched new data if the store was empty (compare [`fetchIfEmpty`](https://github.com/Automattic/wp-calypso/blob/master/client/components/data/followers-data/index.jsx#L67-L86)). I think it'd make sense to discuss whether we want to keep it that way.

tyxla replied:
I think I prefer your approach here; it makes sense to refetch if a component is unmounted and then remounted again if the user navigates back to the same page. It's likely a good idea to use the data layer, though, because it'll do its best to dedupe requests if the mount->unmount->mount dance happens very fast.

flootr committed:
add a bunch of JSDocs

flootr committed:
normalize namin & add more tests

flootr committed:
query-followers: add readme

tyxla commented on client/components/data/query-followers/index.jsx:
```diff
@@ -0,0 +1,50 @@
+		if ( this.props.refresh ) {
+			this.interval = setInterval(
+				() => this.refreshFollowers(),
+				this.props.refreshInterval || REFRESH_INTERVAL
```
YAGNI, too. If we don't need to tweak the `refreshInterval` per use case, we might as well remove it completely.

flootr replied:
TIL YAGNI 😅 – I agree, let's remove it.

flootr replied:
done via b4f1921

tyxla commented on client/components/data/query-followers/index.jsx:
```diff
@@ -0,0 +1,50 @@
+	componentDidMount() {
+		this.props.fetchFollowers( this.props.query );
+
+		if ( this.props.refresh ) {
```
IMHO, YAGNI. If the legacy `followers-data` component didn't have a way to disable polling, why should the query component have it? I'd suggest removing the prop and always enabling polling unless we have a use case where we need to disable it. 

This is also compatible with the [WordPress phisolophy of Decisions, not Options](https://wordpress.org/about/philosophy/#decisions).

flootr replied:
done via b4f1921

tyxla commented on client/components/data/query-followers/README.md:
```diff
@@ -0,0 +1,60 @@
+
+A bunch of query params to filter the API response. See the [API documentation](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/stats/followers/) for a detailed list of options.
+
+### `refresh`
```
If you agree on my suggestions down there to remove those props, let's remove their docs, too.

flootr replied:
done via b4f1921

tyxla commented on client/components/data/query-followers/index.jsx:
```diff
@@ -0,0 +1,50 @@
+ */
+import React from 'react';
+import { connect } from 'react-redux';
+import isShallowEqual from '@wordpress/is-shallow-equal';
```
Nit: We generally keep `@wordpress` imports in their own group, with a `WordPress dependencies` doc block above them.

flootr replied:
TIL 👍🏻  (fixed via 54c8f96)

tyxla commented on client/components/data/query-followers/index.jsx:
```diff
@@ -0,0 +1,50 @@
+
+const REFRESH_INTERVAL = 1000 * 30;
+
+class QueryFollowers extends React.Component {
```
I think we could simplify this by refactoring it to a functional component. This is especially valid if we remove the `refresh*` props. The timeout can be handled in a `useEffect`, just like the `fetchFollowers`.

flootr replied:
We compare the `query` prop with `isShallowEqual` rather than by reference. Is `useEffect` able to account for that? I'm happy for any pointers 👍🏻

tyxla replied:
That's a good point. Let's keep it as it is then.

tyxla commented on client/my-sites/people/followers-list/index.jsx:
```diff
@@ -16,290 +16,286 @@ import PeopleListItem from 'calypso/my-sites/people/people-list-item';
  * Stylesheet dependencies
  */
 import './style.scss';
+import { connect } from 'react-redux';
```
Did we mean to move that up to the `External dependencies` section?

flootr replied:
done via 5b35750

tyxla commented on client/my-sites/people/followers-list/index.jsx:
```diff
@@ -16,290 +16,286 @@ import PeopleListItem from 'calypso/my-sites/people/people-list-item';
-			);
-		}
+		this.props.incrementPage();
+		gaRecordEvent( 'People', analyticsAction, 'page', this.props.currentPage + 1 );
```
Would be cool if we could port those to Redux, too. Not necessarily in this PR, of course.

flootr replied:
I'll follow up with another PR 👍🏻

tyxla commented on client/state/followers/selectors.js:
```diff
@@ -3,21 +3,29 @@
+	);
+};
+
+export const getIsFetchingFollowersForQuery = ( state, query ) => {
```
Naming nit: I'd rename this to `isFetchingFollowersForQuery`; `get` feels pretty redundant in a boolean "is requesting" selector.

flootr replied:
fixed via c9a9322

tyxla commented on client/my-sites/people/followers-list/index.jsx:
```diff
@@ -16,290 +16,286 @@ import PeopleListItem from 'calypso/my-sites/people/people-list-item';
-	} else {
-		DataComponent = FollowersData;
-	}
+const ConnectedFollowers = connect( mapStateToProps, { removeFollower } )( localize( Followers ) );
```
Should we move this component to a separate file? Or should we merge `FollowersList` and `Followers`? In general, what's the idea behind having them as 2 separate components?

flootr replied:
I split them up in d46e1b74a8. I think the reason why these are separate are because the `<Followers />` component renders differently depending on state.

Edit: we can probably refactor that component so that we won't need to have both. If that's wanted I'd like to follow up with another PR instead of doing it in this one (it's already quite huge).

tyxla commented on client/state/followers/actions.js:
```diff
@@ -19,38 +10,82 @@ import {
+ * @param {object} query params which were sent to the API
+ * @param {object} data API response object
+ */
+export const successRequestFollowers = ( query, data ) => ( {
```
Naming nit: I'd rename this to `requestFollowersSuccess`.

flootr replied:
done via 3b3bfc8

tyxla commented on client/state/followers/actions.js:
```diff
@@ -19,38 +10,82 @@ import {
+ * @param {object} query params which were sent to the API
+ * @param {object} error contains the error response
+ */
+export const failRequestFollowers = ( query, error ) => ( {
```
Naming nit: I'd rename this to `requestFollowersFailure` or `requestFollowersError`.

That applies to the rest of the action creators in this file.

flootr replied:
done via 3b3bfc8

tyxla commented on client/state/followers/actions.js:
```diff
@@ -19,38 +10,82 @@ import {
+ * be removed from the site.
+ *
+ * @param {number} siteId site identifier
+ * @param {object} follower follower object
```
```
follower follower object
```

Seems like docs need some tweaking here and there. Sentence case for the description will fix the issue. This applies to multiple areas in this file.

flootr replied:
My editor doesn't complain about those. Does 8de8b7c work?

tyxla replied:
Yep, that does the job for me. My editor didn't complain, either, but my eyes do 😬

tyxla commented on client/state/followers/actions.js:
```diff
@@ -19,38 +10,82 @@ import {
+
+/**
+ * Returns an action object used in signalling that a request for followers
+ * was successfull.
```
```suggestion
 * was successful.
```

flootr replied:
done via 9ab502a

tyxla commented on client/state/followers/selectors.js:
```diff
@@ -3,21 +3,29 @@
+export const getFollowersForQuery = ( state, query ) => {
+	const serializedQuery = getSerializedQuery( query );
+	return (
+		state.followers?.queries[ serializedQuery ]?.ids.map(
```
Is it possible that `followers` state tree won't exist? It seems it's not modular at this point? Or did you intend to make it modular?

flootr replied:
See my comment https://github.com/Automattic/wp-calypso/pull/47451#discussion_r526917994, will make this modular as part of this PR.

tyxla commented on client/state/followers/test/selectors.js:
```diff
@@ -57,7 +49,49 @@ describe( 'selctors', () => {
+		} );
+	} );
+
+	describe( '#getIsFetchingFollowersForQuery', () => {
```
Do we also need a test to cover the case when a query request with that serialized query doesn't exist at all?

flootr replied:
fixed via 31043440f5

tyxla commented on client/state/followers/test/selectors.js:
```diff
@@ -57,7 +49,49 @@ describe( 'selctors', () => {
+	} );
+
+	describe( '#getTotalFollowersForQuery', () => {
+		test( 'should return the total number of followers for a query', () => {
```
Do we need a test to cover the case when there is no data for that query?

flootr replied:
fixed via 31043440f5

tyxla commented on client/state/followers/thunks/fetch-followers.js:
```diff
@@ -0,0 +1,36 @@
+ * @param {object} query contains params to filter results
+ * @param {boolean} silentUpdate indicates whether the request should be reflected in the UI
+ */
+function fetchFollowers( query, silentUpdate = false ) {
```
I'd love to see this migrated to the data layer. It seems straightforward enough to be directly ported there.

flootr replied:
done via d056358

tyxla commented on client/state/followers/thunks/fetch-followers.js:
```diff
@@ -0,0 +1,36 @@
+ */
+function fetchFollowers( query, silentUpdate = false ) {
+	return ( dispatch ) => {
+		// TODO: Componentes should not fetch if already fetching
```
Ditto. Data layer is a good way to fix this. 😉

flootr replied:
done via d056358

tyxla commented on client/state/followers/thunks/remove-follower.js:
```diff
@@ -0,0 +1,34 @@
+ * @param {number} siteId site identifier
+ * @param {object} follower contains the follower object
+ */
+function removeFollower( siteId, follower ) {
```
Another good candidate for porting to the data layer instead of creating a new thunk.

flootr replied:
done via d056358

tyxla commented on client/state/reducer.js:
```diff
@@ -51,6 +52,7 @@ const reducers = {
 	dataRequests,
 	documentHead,
 	experiments,
+	followers,
```
How does it sound to make the `followers` reducer modular from the beginning? @sgomes would love that 😉

sgomes replied:
I'm happy to help make it happen, too 🙂 [This should help you get started](https://github.com/Automattic/wp-calypso/blob/master/docs/modularized-state.md), but let me know if you have any questions or want any help!

flootr replied:
Yup, I thought about that as well. Will fix that 👍🏻

flootr replied:
done via 838eb7e

tyxla reviewed (commented):
This is a big one 😅 

I've left some initial feedback - I hope that's helpful. Let me know what you think.

I'm also seeing this when I visit `/people/followers/:site`:

![](https://cldup.com/-iyeGbuVcx.png)

jsnajdr commented:
> I'm also seeing this when I visit `/people/followers/:site`:

This can happen when you dispatch an action directly in the `render` method or in the `mapStateToProps` function. Both are supposed to be pure, without side effects.

The React warning is very new ([16.3.0, February 2020](https://reactjs.org/blog/2020/02/26/react-v16.13.0.html#warnings-for-some-updates-during-render)) and it's possible the affected code was already buggy for years.

`@wordpress/data` also had issues with this, when calling a selector immediately fired an associated resolver.

flootr committed:
people(followers-list): add comment for why we're using `key`

flootr committed:
query-followers: make `refresh` mandatory

flootr committed:
cleanup names for followers/actions

flootr committed:
followers-list: move `connect` import to external dependencies

flootr committed:
fix type in followers/list

flootr committed:
query-followers: rearrange imports

flootr committed:
rename getIsFetchingFollowersForQuery to isFetchingFollowersForQuery

flootr committed:
improve tests for state/followers/selectors

flootr committed:
followers/actions: use sentence case for JSDoc descriptions

flootr commented:
> > I'm also seeing this when I visit `/people/followers/:site`:
> 
> This can happen when you dispatch an action directly in the `render` method or in the `mapStateToProps` function. Both are supposed to be pure, without side effects.

I'm seeing this error on the `trunk` branch as well so it's likely not caused by changes in this PR. However, I'll dedicate some time to see if I can fix it.

flootr committed:
modularise followers state

flootr committed:
state(followers): use data-layer

flootr committed:
followers-list: split components in separte files

flootr commented:
@tyxla @jsnajdr I think this one is ready for another round of reviews 👍🏻 I addressed the feedback and/or left comments where appropriate. The state is now modularised and we use the data layer for fetching and removing followers.

sgomes commented:
Thank you for modularising state, @flootr! 👍 We'll get rid of that root reducer yet! 😄

tyxla commented on client/my-sites/people/followers-list/followers.jsx:
```diff
@@ -0,0 +1,270 @@
+	}
+}
+
+const mapStateToProps = ( state, ownProps ) => {
+	return {
+		followers: getFollowersForQuery( state, ownProps.query ),
+		fetching: isFetchingFollowersForQuery( state, ownProps.query ),
+		totalFollowers: getTotalFollowersForQuery( state, ownProps.query ),
+	};
+};
```
Nit: this could be simplified to:

```suggestion
const mapStateToProps = ( state, { query } ) => ( {
	followers: getFollowersForQuery( state, query ),
	fetching: isFetchingFollowersForQuery( state, query ),
	totalFollowers: getTotalFollowersForQuery( state, query ),
} );
```

tyxla commented on client/my-sites/people/followers-list/followers.jsx:
```diff
@@ -0,0 +1,270 @@
+			followers = this.renderPlaceholders();
+		}
+
+		const downloadListLink =
```
Nit: would be cool if we could use a more intelligent way to build a query string for this URL. We have `addQueryArgs()` for example.

Feel free to ignore; I know this was like this before this PR.

tyxla commented on client/my-sites/people/main.jsx:
```diff
@@ -33,9 +33,20 @@ class People extends React.Component {
+						key={ `people-email-followers-${ site.ID }` }
+						site={ site }
+						search={ search }
+						type="email"
```
Uber nit: seeing a `type` prop outside an input file hurts my eyes, literally. 

Don't feel obligated to change that, not in this PR anyway. I just wanted to share this 😉

tyxla commented on client/state/followers/actions.js:
```diff
@@ -19,38 +10,87 @@ import {
+
+/**
+ * Returns an action object used in signalling that followers for the site
+ * have been requested..
```
```suggestion
 * have been requested.
```

tyxla commented on client/state/followers/init.js:
```diff
@@ -0,0 +1,7 @@
+/**
```
Thanks for modularizing it ❤️

tyxla commented on client/state/followers/test/selectors.js:
```diff
@@ -1,12 +1,11 @@
+} from '../selectors';
 import { getSerializedQuery } from '../utils';
 
 describe( 'selctors', () => {
```
There's a pre-existing typo here:

```suggestion
describe( 'selectors', () => {
```

If you ask me, this `describe` is unnecessary, it's clear that this is testing selectors when seeing the name of this file.

tyxla commented on client/state/followers/test/selectors.js:
```diff
@@ -57,7 +49,61 @@ describe( 'selctors', () => {
 			};
 			const query = { siteId: 555 };
-			expect( getFollowersForQuery( state, query ) ).to.be.null;
+			expect( getFollowersForQuery( state, query ) ).toBeNull;
```
Did you mean to invoke it?
```suggestion
			expect( getFollowersForQuery( state, query ) ).toBeNull();
```

tyxla commented on client/state/followers/reducer.js:
```diff
@@ -80,9 +82,12 @@ export function queryRequests( state = {}, action ) {
 			const serializedQuery = getSerializedQuery( action.query );
-			return Object.assign( {}, state, { [ serializedQuery ]: FOLLOWERS_REQUEST === action.type } );
+			return Object.assign( {}, state, {
+				[ serializedQuery ]: FOLLOWERS_REQUEST === action.type && ! action.silentUpdate,
```
Should this change be reflected in the reducer tests?

tyxla commented on client/state/data-layer/wpcom/followers/index.js:
```diff
@@ -0,0 +1,73 @@
+		http(
+			{
+				method: 'POST',
+				path: `sites/${ siteId }/followers/${ follower.ID }/delete`,
```
Removing followers currently won't work because this `path` is missing a leading `/`.

tyxla reviewed (changes requested):
I've tested with a site with 7K+ followers and 5K+ email followers, everything works great! 

The only thing that doesn't quite work is removing followers - I'm seeing а 404 when attempting to remove a follower:

![](https://cldup.com/EiiVNkOE9B.png)

I've also left some nitpicks, nothing serious.

Nice work 👏

flootr committed:
fix followers selectors tests

flootr committed:
add tests for followers.queryRequests reducer

flootr committed:
migrate followers reducer tests to jest

flootr committed:
removeFollowers: add leading slash to API url

flootr committed:
followers: use `addQueryArgs` to build csv download link

flootr committed:
fix typo in followers/actions.js

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

flootr committed:
simplify mapStateToProps in followers-list/followers.jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

flootr commented:
@tyxla removing followers should now work per 044ec31, would you mind having another look? 🙏🏻

tyxla reviewed (approved):
Looks and works great, 🚢  it!

Awesome work @flootr 🙌
